File: SessionTimeFinder.java
package com.asha.test;

import java.io.IOException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.hadoop.io.Text;
import org.apache.hadoop.mapreduce.Job;
import org.apache.hadoop.mapreduce.Mapper;
import org.apache.hadoop.mapreduce.Reducer;
import org.apache.hadoop.mapreduce.lib.input.FileInputFormat;
import org.apache.hadoop.mapreduce.lib.output.FileOutputFormat;

/**
 * The class finds the latest visiting time for each user/ip in a session. 
 * Assume:
 *   - session time window is fixed, 15 minutes
 *   - this session time window starts at 2015-07-22T09:00:00.000Z
 *   - this session time windows ends at 2015-07-22T09:15:00.000Z
 * For example, if user1 visited some url at 2015-07-22T09:01:00.000Z and 2015-07-22T09:02:00.000Z,
 * the output for this user is:  (user1  2015-07-22T09:02:00.000Z).
 * It runs as a mapreduce job on a hadoop cluster.
 * Mapper prepares the data as (IP, Visiting time)
 * Reducer finds the most recent visiting time for each ip.
 * @author asha
 *
 */
public class SessionTimeFinder {

	public static class GetSessionTimeMapper extends Mapper<Object, Text, Text, Text> {

		private Text ip = new Text();
		private Text time = new Text();
		private static SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");

		public void map(Object key, Text value, Context context) throws IOException, InterruptedException {
			try {
				String[] results = value.toString().split("\\s");
				String visitTime = results[0];
				String ipStr = results[2];
				Date startTime = sdf.parse("2015-07-22T09:00:00.000Z");
				Date endTime = sdf.parse("2015-07-22T09:15:00.000Z");
				String currentTimeStr = visitTime.substring(0, 20)
						+ String.valueOf(Integer.parseInt(visitTime.substring(20, 26)) / 1000) + "Z";
				Date currentTime = sdf.parse(currentTimeStr);
				if (currentTime.after(startTime) && currentTime.before(endTime)) {
					ip.set(ipStr);
					time.set(currentTimeStr);
					context.write(ip, time);
				}
			} catch (ParseException e) {
				System.out.println("Parsing error for the timestamp.");
			}
		}
	}

	public static class GetSessionTimeReducer extends Reducer<Text, Text, Text, Text> {
		private static SimpleDateFormat sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");

		public void reduce(Text key, Iterable<Text> values, Context context) throws IOException, InterruptedException {
			try {
				Date startTime = sdf.parse("2015-07-22T09:00:00.000Z");
				Text maxTime = new Text("2015-07-22T09:00:00.000Z");
				for (Text val : values) {
					Date currentTime = sdf.parse(val.toString());
					if (currentTime.after(startTime)) {
						startTime = currentTime;
						maxTime = val;
					}
				}
				context.write(key, maxTime);
			} catch (ParseException e) {
				System.out.println("Parsing error for the timestamp.");
			}
		}
	}

	public static void main(String[] args) throws Exception {
		Configuration conf = new Configuration();
		Job job = Job.getInstance(conf, "session time");
		job.setJarByClass(SessionTimeFinder.class);
		job.setMapperClass(GetSessionTimeMapper.class);
		job.setCombinerClass(GetSessionTimeReducer.class);
		job.setReducerClass(GetSessionTimeReducer.class);
		job.setOutputKeyClass(Text.class);
		job.setOutputValueClass(Text.class);
		FileInputFormat.addInputPath(job, new Path(args[0]));
		FileOutputFormat.setOutputPath(job, new Path(args[1]));
		System.exit(job.waitForCompletion(true) ? 0 : 1);
	}
}

File: scala/src/main/scala/SessionTimeFinder.scala
import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.apache.spark.SparkConf

import java.io._
import java.lang.Math
import java.text.SimpleDateFormat

/**
 * This class aggregates the number of the visits for each user. (goal 1 in the task list)
 * Assume:
 *   - session time window is fixed, 15 minutes
 *   - this session time window starts at 2015-07-22T09:00:00.000Z
 *   - this session time windows ends at 2015-07-22T09:15:00.000Z
 *   - same visitorip but different port considered to be different visitor
 * It runs as a mapreduce job on a hadoop cluster.
 * Mapper prepares the data as (IP, 1)
 * Reducer adds the number up for each IP.
 * @author asha
 *
 */
object AverageSessionTime {

	def withinTime(line: String) : Boolean = {
		//do something
		val words = line.split("\\s")
		val visitTime = words(0)
		val sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
		val startTime = sdf.parse("2015-07-22T09:00:00.000Z")
		val endTime = sdf.parse("2015-07-22T09:15:00.000Z")
		val currentTimeStr = visitTime.substring(0, 20) + String.valueOf(Integer.parseInt(visitTime.substring(20, 26)) / 1000) + "Z"
		val currentTime = sdf.parse(currentTimeStr)

		if (currentTime.after(startTime) && currentTime.before(endTime)) {
			return true
		}

		return false
	}

	def getVisitorIpAndTime(line: String) : (String, String) = {
		val words = line.split("\\s")
		val visitTime = words(0)
		val visitorIp = words(2)
		val sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
		val currentTimeStr = visitTime.substring(0, 20) + String.valueOf(Integer.parseInt(visitTime.substring(20, 26)) / 1000) + "Z";

		return (visitorIp, currentTimeStr)
	}

	def getMaxTime(timeA:String, timeB:String) : String = {
		val sdf = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
		val time1 = sdf.parse(timeA)
		val time2 = sdf.parse(timeB)

		if (time1.after(time2)) return timeA
		
		return timeB
	}

	def main(args: Array[String]) : Unit = {

		val conf = new SparkConf().setAppName("SessionTimeFinder")
		val sc = new SparkContext(conf)

		val inputLog = sc.textFile("data/2015_07_22_mktplace_shop_web_log_sample.log")

		val hitsCounts = inputLog.filter(line => withinTime(line)).map(line => (getVisitorIpAndTime(line))).reduceByKey((a, b) => getMaxTime(a, b))

		hitsCounts.saveAsTextFile("avgSessionTime")
		
		val output = new PrintWriter(new File("sessionTime.out"))
		hitsCounts.collect().foreach(line => output.println(line)) 
		output.close()
	}
}
